modules/compute/src/main/java/org/apache/ignite/compute/internal/IgniteComputeImpl.java

```diff
@@ -0,0 +1,302 @@
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.apache.ignite.compute.ComputeException;
+import org.apache.ignite.compute.ComputeTask;
+import org.apache.ignite.compute.IgniteCompute;
+import org.apache.ignite.compute.internal.message.ComputeMessagesFactory;
+import org.apache.ignite.compute.internal.message.ComputeMessagesType;
+import org.apache.ignite.compute.internal.message.ExecuteMessage;
+import org.apache.ignite.compute.internal.message.ResultMessage;
+import org.apache.ignite.internal.manager.IgniteComponent;
+import org.apache.ignite.internal.util.IgniteUtils;
+import org.apache.ignite.network.ClusterNode;
+import org.apache.ignite.network.ClusterService;
+import org.apache.ignite.network.MessagingService;
+import org.apache.ignite.network.NetworkAddress;
+import org.apache.ignite.network.NetworkMessage;
+import org.apache.ignite.network.NetworkMessageHandler;
+import org.apache.ignite.network.TopologyEventHandler;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Implementation of the compute service.
+ */
+public class IgniteComputeImpl implements IgniteCompute, IgniteComponent {
+    private final ClusterService clusterService;
+
+    private final ExecutorService computeExecutor = Executors.newSingleThreadExecutor();
+
+    private final ComputeMessagesFactory messagesFactory = new ComputeMessagesFactory();
+
+    private final ConcurrentMap<UUID, Execution<?>> executions = new ConcurrentHashMap<>();
+
+    public IgniteComputeImpl(ClusterService clusterService) {
+        this.clusterService = clusterService;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public <T, R, F> F execute(Class<? extends ComputeTask<T, R, F>> clazz, T argument) throws ComputeException {
+        UUID executionId = UUID.randomUUID();
+
+        Execution<R> execution = registerExecution(clazz, argument, executionId);
+
+        Collection<ExecutionResult<R>> executionResults = execution.getResult();
+
+        executions.remove(executionId);
+
+        ComputeTask<T, R, F> computeTask;
+        try {
+            Constructor<? extends ComputeTask<T, R, F>> declaredConstructor = clazz.getDeclaredConstructor();
+            declaredConstructor.setAccessible(true);
+            computeTask = declaredConstructor.newInstance();
+        } catch (InstantiationException | IllegalAccessException | NoSuchMethodException | InvocationTargetException e) {
+            throw new ComputeException("Failed to execute reduce: " + e.getMessage(), e);
+        }
+
+        return computeTask.reduce(executionResults.stream().map(ExecutionResult::result).collect(toList()));
```

Comment:
Is it ok that `reduce()` is executed on the caller thread?

Reply:
I think so, yes. No benefits from sending it to another thread and waiting in the current thread, I suppose, only performance downsides